File: android/app/src/main/java/com/learnpaper/render/CardRenderer.kt
package com.learnpaper.render

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BlurMaskFilter
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import android.text.TextUtils
import com.learnpaper.content.ContentRepository
import com.learnpaper.content.Lang
import com.learnpaper.content.Word
import com.learnpaper.data.LayoutPreset
import com.learnpaper.data.Settings
import kotlin.math.max

/**
 * Draws a word card onto a bitmap of the given size.
 *
 * All dimensions are expressed for a 1080 px wide canvas and scaled by `width / 1080`.
 * The content is a vertical stack of blocks centred inside a band that depends on the
 * layout preset (lower part of the screen for the lock screen, so the clock stays clear).
 * If the stack does not fit, everything is shrunk step by step.
 */
class CardRenderer(private val context: Context, private val content: ContentRepository) {

    fun render(word: Word, settings: Settings, palette: Palette, width: Int, height: Int): Bitmap {
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val s = width / 1080f
        drawBackground(canvas, palette, width.toFloat(), height.toFloat(), s)

        val (bandTop, bandBottom) = when (settings.layout) {
            LayoutPreset.LOCK -> 0.30f to 0.90f
            LayoutPreset.HOME -> 0.16f to 0.88f
            LayoutPreset.COMPACT -> 0.58f to 0.92f
        }
        val top = bandTop * height
        val bottom = bandBottom * height
        val maxWidth = width - 2 * MARGIN * s
        val image = if (settings.layout == LayoutPreset.COMPACT) null else content.loadImage(word)

        var scale = 1f
        var blocks = buildBlocks(word, settings, palette, image, s, maxWidth)
        var total = blocks.sumOf { it.height.toDouble() }.toFloat()
        while (total > bottom - top && scale > 0.5f) {
            scale *= 0.92f
            blocks = buildBlocks(word, settings, palette, image, s * scale, maxWidth)
            total = blocks.sumOf { it.height.toDouble() }.toFloat()
        }

        var y = top + ((bottom - top) - total) / 2f
        val cx = width / 2f
        for (block in blocks) {
            block.draw(canvas, cx, y)
            y += block.height
        }
        image?.recycle()
        return bitmap
    }

    private fun drawBackground(canvas: Canvas, palette: Palette, w: Float, h: Float, s: Float) {
        canvas.drawColor(palette.bg)
        val deco = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = palette.tile
            maskFilter = BlurMaskFilter(140f * s, BlurMaskFilter.Blur.NORMAL)
        }
        canvas.drawCircle(w * 0.88f, h * 0.10f, 300f * s, deco)
        canvas.drawCircle(w * 0.10f, h * 0.96f, 340f * s, deco)
    }

    private fun buildBlocks(
        word: Word,
        settings: Settings,
        palette: Palette,
        image: Bitmap?,
        s: Float,
        maxWidth: Float,
    ): List<Block> {
        val blocks = ArrayList<Block>()
        val headline = word.entry(settings.headline)
        val translations = settings.translations
        val compact = settings.layout == LayoutPreset.COMPACT

        if (image != null) {
            blocks += ImageTile(image, size = 360f * s, radius = 64f * s, tileColor = palette.tile)
            blocks += Spacer(56f * s)
        }

        val headlineSize = fitSize(headline.text, 88f * s, 700, maxWidth, minSize = 40f * s)
        blocks += TextLine(headline.text, paint(headlineSize, 700, palette.text))
        if (settings.showTranscriptions && headline.tr.isNotBlank()) {
            blocks += Spacer(10f * s)
            blocks += TextLine(formatTr(settings.headline, headline.tr), paint(34f * s, 500, palette.muted))
        }
        blocks += Spacer(20f * s)
        val tag = listOf(word.level, PosNames.localized(context, word.pos)).filter { it.isNotBlank() }
            .joinToString("  ·  ").uppercase()
        blocks += Tag(tag, paint(22f * s, 600, palette.muted).apply { letterSpacing = 0.12f }, palette.tile, s)

        if (translations.isNotEmpty()) {
            blocks += Spacer(if (compact) 44f * s else 56f * s)
            translations.forEachIndexed { i, lang ->
                if (i > 0) blocks += Spacer(22f * s)
                val e = word.entry(lang)
                blocks += WordLine(
                    label = lang.label,
                    labelPaint = paint(22f * s, 700, palette.muted).apply { letterSpacing = 0.1f },
                    labelBg = palette.tile,
                    text = e.text,
                    textPaint = paint(fitSize(e.text, 46f * s, 600, maxWidth * 0.7f, minSize = 28f * s), 600, palette.text),
                    tr = if (settings.showTranscriptions && e.tr.isNotBlank()) formatTr(lang, e.tr) else null,
                    trPaint = paint(30f * s, 500, palette.muted),
                    gap = 18f * s,
                    maxWidth = maxWidth,
                    s = s,
                )
            }
        }

        if (settings.showExamples && !compact) {
            val main = word.example.of(settings.headline)
            if (main.isNotBlank()) {
                blocks += Spacer(48f * s)
                blocks += Divider(300f * s, 2f * s, withAlpha(palette.muted, 0.35f))
                blocks += Spacer(40f * s)
                blocks += Paragraph(main, paint(34f * s, 500, palette.text), maxWidth, 1.2f)
                for (lang in translations) {
                    val t = word.example.of(lang)
                    if (t.isBlank()) continue
                    blocks += Spacer(16f * s)
                    blocks += Paragraph(t, paint(28f * s, 400, palette.muted), maxWidth, 1.2f)
                }
            }
        }
        return blocks
    }

    private fun paint(size: Float, weight: Int, color: Int): TextPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
        typeface = Fonts.get(context, weight)
        textSize = size
        this.color = color
        isSubpixelText = true
    }

    private fun fitSize(text: String, start: Float, weight: Int, maxWidth: Float, minSize: Float): Float {
        var size = start
        val p = paint(size, weight, 0)
        while (size > minSize && p.measureText(text) > maxWidth) {
            size *= 0.94f
            p.textSize = size
        }
        return size
    }

    private fun formatTr(lang: Lang, tr: String): String = if (lang == Lang.EN) "/$tr/" else "[$tr]"

    private fun withAlpha(color: Int, alpha: Float): Int =
        (color and 0x00FFFFFF) or ((alpha * 255).toInt().coerceIn(0, 255) shl 24)

    // --- blocks -----------------------------------------------------------------

    private interface Block {
        val height: Float
        fun draw(canvas: Canvas, cx: Float, y: Float)
    }

    private class Spacer(override val height: Float) : Block {
        override fun draw(canvas: Canvas, cx: Float, y: Float) = Unit
    }

    private class TextLine(private val text: String, private val paint: TextPaint) : Block {
        private val fm = paint.fontMetrics
        override val height = fm.descent - fm.ascent
        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            canvas.drawText(text, cx - paint.measureText(text) / 2f, y - fm.ascent, paint)
        }
    }

    /** Small uppercase pill, e.g. "A1 · NOUN". */
    private class Tag(private val text: String, private val paint: TextPaint, private val bg: Int, s: Float) : Block {
        private val fm = paint.fontMetrics
        private val padX = 22f * s
        private val padY = 10f * s
        override val height = fm.descent - fm.ascent + 2 * padY
        private val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = bg }
        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            val textW = paint.measureText(text)
            val w = textW + 2 * padX
            canvas.drawRoundRect(RectF(cx - w / 2f, y, cx + w / 2f, y + height), height / 2f, height / 2f, fill)
            canvas.drawText(text, cx - textW / 2f, y + padY - fm.ascent, paint)
        }
    }

    /** "[RU] зонт [zont]" on one line, with the transcription wrapping below when it does not fit. */
    private class WordLine(
        private val label: String,
        private val labelPaint: TextPaint,
        labelBg: Int,
        private val text: String,
        private val textPaint: TextPaint,
        private val tr: String?,
        private val trPaint: TextPaint,
        private val gap: Float,
        maxWidth: Float,
        s: Float,
    ) : Block {
        private val labelPadX = 14f * s
        private val labelPadY = 8f * s
        private val labelFm = labelPaint.fontMetrics
        private val labelH = labelFm.descent - labelFm.ascent + 2 * labelPadY
        private val labelW = labelPaint.measureText(label) + 2 * labelPadX
        private val textW = textPaint.measureText(text)
        private val trW = tr?.let { trPaint.measureText(it) } ?: 0f
        private val oneLine = tr == null || labelW + gap + textW + gap + trW <= maxWidth
        private val textFm = textPaint.fontMetrics
        private val trFm = trPaint.fontMetrics
        private val ascent = max(-textFm.ascent, if (oneLine && tr != null) -trFm.ascent else 0f)
        private val descent = max(textFm.descent, if (oneLine && tr != null) trFm.descent else 0f)
        private val line1H = max(ascent + descent, labelH)
        private val line2Gap = 6f * s
        private val line2H = if (!oneLine && tr != null) (trFm.descent - trFm.ascent) + line2Gap else 0f
        private val pill = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = labelBg }
        override val height = line1H + line2H

        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            val total = labelW + gap + textW + if (oneLine && tr != null) gap + trW else 0f
            var x = cx - total / 2f
            val pillTop = y + (line1H - labelH) / 2f
            canvas.drawRoundRect(RectF(x, pillTop, x + labelW, pillTop + labelH), labelH / 2f, labelH / 2f, pill)
            canvas.drawText(label, x + labelPadX, pillTop + labelPadY - labelFm.ascent, labelPaint)
            x += labelW + gap
            val baseline = y + (line1H - (ascent + descent)) / 2f + ascent
            canvas.drawText(text, x, baseline, textPaint)
            if (tr != null) {
                if (oneLine) {
                    canvas.drawText(tr, x + textW + gap, baseline, trPaint)
                } else {
                    canvas.drawText(tr, cx - trW / 2f, y + line1H + line2Gap - trFm.ascent, trPaint)
                }
            }
        }
    }

    private class Paragraph(text: String, paint: TextPaint, private val width: Float, lineSpacing: Float) : Block {
        private val layout: StaticLayout = StaticLayout.Builder.obtain(text, 0, text.length, paint, width.toInt())
            .setAlignment(Layout.Alignment.ALIGN_CENTER)
            .setLineSpacing(0f, lineSpacing)
            .setIncludePad(false)
            .setMaxLines(3)
            .setEllipsize(TextUtils.TruncateAt.END)
            .build()
        override val height = layout.height.toFloat()
        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            canvas.save()
            canvas.translate(cx - width / 2f, y)
            layout.draw(canvas)
            canvas.restore()
        }
    }

    private class Divider(private val w: Float, thickness: Float, color: Int) : Block {
        override val height = thickness
        private val paint = Paint().apply { this.color = color }
        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            canvas.drawRect(cx - w / 2f, y, cx + w / 2f, y + height, paint)
        }
    }

    private class ImageTile(
        private val bitmap: Bitmap,
        private val size: Float,
        private val radius: Float,
        tileColor: Int,
    ) : Block {
        override val height = size
        private val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = tileColor }
        private val img = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)
        override fun draw(canvas: Canvas, cx: Float, y: Float) {
            canvas.drawRoundRect(RectF(cx - size / 2f, y, cx + size / 2f, y + size), radius, radius, fill)
            val inner = size * 0.64f
            val dst = RectF(cx - inner / 2f, y + (size - inner) / 2f, cx + inner / 2f, y + (size + inner) / 2f)
            canvas.drawBitmap(bitmap, null, dst, img)
        }
    }

    companion object {
        private const val MARGIN = 96f
    }
}

File: ios/Shared/CardRenderer.swift
import UIKit

/// Draws a word card into an image of the given pixel size. Port of the Android `CardRenderer`:
/// all dimensions are expressed for a 1080 px wide canvas and scaled by `width / 1080`; the
/// content is a vertical stack of blocks centred inside a band that depends on the layout preset,
/// shrunk step by step if it does not fit.
final class CardRenderer {
    static let shared = CardRenderer()
    private let margin: CGFloat = 96

    func render(word: Word, settings: Settings, palette: Palette, width: CGFloat, height: CGFloat) -> UIImage {
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        let size = CGSize(width: width, height: height)
        return UIGraphicsImageRenderer(size: size, format: format).image { rc in
            let ctx = rc.cgContext
            let s = width / 1080
            drawBackground(ctx, palette: palette, w: width, h: height, s: s)

            let (bandTop, bandBottom): (CGFloat, CGFloat)
            switch settings.layout {
            case .lock: (bandTop, bandBottom) = (0.30, 0.90)
            case .home: (bandTop, bandBottom) = (0.16, 0.88)
            case .compact: (bandTop, bandBottom) = (0.58, 0.92)
            }
            let top = bandTop * height
            let bottom = bandBottom * height
            let maxWidth = width - 2 * margin * s
            let image = settings.layout == .compact ? nil : ContentStore.shared.image(for: word)

            var scale: CGFloat = 1
            var blocks = buildBlocks(word: word, settings: settings, palette: palette, image: image, s: s, maxWidth: maxWidth)
            var total = blocks.reduce(0) { $0 + $1.height }
            while total > bottom - top && scale > 0.5 {
                scale *= 0.92
                blocks = buildBlocks(word: word, settings: settings, palette: palette, image: image, s: s * scale, maxWidth: maxWidth)
                total = blocks.reduce(0) { $0 + $1.height }
            }

            var y = top + ((bottom - top) - total) / 2
            let cx = width / 2
            for block in blocks {
                block.draw(cx: cx, y: y, ctx: ctx)
                y += block.height
            }
        }
    }

    private func drawBackground(_ ctx: CGContext, palette: Palette, w: CGFloat, h: CGFloat, s: CGFloat) {
        ctx.setFillColor(palette.bg.cgColor)
        ctx.fill(CGRect(x: 0, y: 0, width: w, height: h))
        // Soft blobs, like the blurred circles on Android.
        for (center, radius) in [(CGPoint(x: w * 0.88, y: h * 0.10), 300 * s), (CGPoint(x: w * 0.10, y: h * 0.96), 340 * s)] {
            let colors = [palette.tile.cgColor, palette.tile.withAlphaComponent(0).cgColor] as CFArray
            guard let gradient = CGGradient(colorsSpace: CGColorSpaceCreateDeviceRGB(), colors: colors, locations: [0.35, 1]) else { continue }
            ctx.drawRadialGradient(gradient, startCenter: center, startRadius: 0, endCenter: center, endRadius: radius * 1.5, options: [])
        }
    }

    private func buildBlocks(word: Word, settings: Settings, palette: Palette, image: UIImage?, s: CGFloat, maxWidth: CGFloat) -> [Block] {
        var blocks: [Block] = []
        let headline = word.entry(settings.headline)
        let translations = settings.translations
        let compact = settings.layout == .compact

        if let image {
            blocks.append(ImageTile(image: image, size: 360 * s, radius: 64 * s, tileColor: palette.tile))
            blocks.append(Spacer(56 * s))
        }

        let headlineSize = fitSize(headline.text, start: 88 * s, weight: 700, maxWidth: maxWidth, minSize: 40 * s)
        blocks.append(TextLine(headline.text, attrs(headlineSize, 700, palette.text)))
        if settings.showTranscriptions, !headline.tr.isEmpty {
            blocks.append(Spacer(10 * s))
            blocks.append(TextLine(Self.formatTr(settings.headline, headline.tr), attrs(34 * s, 500, palette.muted)))
        }
        blocks.append(Spacer(20 * s))
        let tag = [word.level, PosNames.localized(word.pos)].filter { !$0.isEmpty }.joined(separator: "  ·  ").uppercased()
        blocks.append(Tag(tag, attrs(22 * s, 600, palette.muted, kern: 0.12 * 22 * s), bg: palette.tile, s: s))

        if !translations.isEmpty {
            blocks.append(Spacer((compact ? 44 : 56) * s))
            for (i, lang) in translations.enumerated() {
                if i > 0 { blocks.append(Spacer(22 * s)) }
                let e = word.entry(lang)
                blocks.append(WordLine(
                    label: lang.label,
                    labelAttrs: attrs(22 * s, 700, palette.muted, kern: 0.1 * 22 * s),
                    labelBg: palette.tile,
                    text: e.text,
                    textAttrs: attrs(fitSize(e.text, start: 46 * s, weight: 600, maxWidth: maxWidth * 0.7, minSize: 28 * s), 600, palette.text),
                    tr: settings.showTranscriptions && !e.tr.isEmpty ? Self.formatTr(lang, e.tr) : nil,
                    trAttrs: attrs(30 * s, 500, palette.muted),
                    gap: 18 * s,
                    maxWidth: maxWidth,
                    s: s
                ))
            }
        }

        if settings.showExamples, !compact {
            let main = word.example.of(settings.headline)
            if !main.isEmpty {
                blocks.append(Spacer(48 * s))
                blocks.append(Divider(width: 300 * s, thickness: 2 * s, color: palette.muted.withAlphaComponent(0.35)))
                blocks.append(Spacer(40 * s))
                blocks.append(Paragraph(main, attrs(34 * s, 500, palette.text), width: maxWidth, lineSpacing: 1.2))
                for lang in translations {
                    let t = word.example.of(lang)
                    if t.isEmpty { continue }
                    blocks.append(Spacer(16 * s))
                    blocks.append(Paragraph(t, attrs(28 * s, 400, palette.muted), width: maxWidth, lineSpacing: 1.2))
                }
            }
        }
        return blocks
    }

    private func attrs(_ size: CGFloat, _ weight: Int, _ color: UIColor, kern: CGFloat = 0) -> [NSAttributedString.Key: Any] {
        var a: [NSAttributedString.Key: Any] = [.font: Fonts.inter(size: size, weight: weight), .foregroundColor: color]
        if kern != 0 { a[.kern] = kern }
        return a
    }

    private func fitSize(_ text: String, start: CGFloat, weight: Int, maxWidth: CGFloat, minSize: CGFloat) -> CGFloat {
        var size = start
        while size > minSize, (text as NSString).size(withAttributes: [.font: Fonts.inter(size: size, weight: weight)]).width > maxWidth {
            size *= 0.94
        }
        return size
    }

    static func formatTr(_ lang: Lang, _ tr: String) -> String { lang == .en ? "/\(tr)/" : "[\(tr)]" }

    // MARK: - blocks

    private struct Metrics {
        let ascent: CGFloat
        let descent: CGFloat
        var height: CGFloat { ascent + descent }
        init(_ attrs: [NSAttributedString.Key: Any]) {
            let font = (attrs[.font] as? UIFont) ?? UIFont.systemFont(ofSize: 16)
            ascent = font.ascender
            descent = -font.descender
        }
    }

    private static func measure(_ text: String, _ attrs: [NSAttributedString.Key: Any]) -> CGFloat {
        (text as NSString).size(withAttributes: attrs).width
    }

    private static func drawText(_ text: String, _ attrs: [NSAttributedString.Key: Any], x: CGFloat, baseline: CGFloat, ascent: CGFloat) {
        (text as NSString).draw(at: CGPoint(x: x, y: baseline - ascent), withAttributes: attrs)
    }

    private struct Spacer: Block {
        let height: CGFloat
        init(_ h: CGFloat) { height = h }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {}
    }

    private struct TextLine: Block {
        let text: String
        let attrs: [NSAttributedString.Key: Any]
        let m: Metrics
        let height: CGFloat
        init(_ text: String, _ attrs: [NSAttributedString.Key: Any]) {
            self.text = text
            self.attrs = attrs
            m = Metrics(attrs)
            height = m.height
        }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            CardRenderer.drawText(text, attrs, x: cx - CardRenderer.measure(text, attrs) / 2, baseline: y + m.ascent, ascent: m.ascent)
        }
    }

    /// Small uppercase pill, e.g. "A1 · NOUN".
    private struct Tag: Block {
        let text: String
        let attrs: [NSAttributedString.Key: Any]
        let bg: UIColor
        let padX: CGFloat
        let padY: CGFloat
        let m: Metrics
        let height: CGFloat
        init(_ text: String, _ attrs: [NSAttributedString.Key: Any], bg: UIColor, s: CGFloat) {
            self.text = text
            self.attrs = attrs
            self.bg = bg
            padX = 22 * s
            padY = 10 * s
            m = Metrics(attrs)
            height = m.height + 2 * padY
        }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            let textW = CardRenderer.measure(text, attrs)
            let w = textW + 2 * padX
            ctx.setFillColor(bg.cgColor)
            ctx.addPath(UIBezierPath(roundedRect: CGRect(x: cx - w / 2, y: y, width: w, height: height), cornerRadius: height / 2).cgPath)
            ctx.fillPath()
            CardRenderer.drawText(text, attrs, x: cx - textW / 2, baseline: y + padY + m.ascent, ascent: m.ascent)
        }
    }

    /// "[RU] зонт [zont]" on one line, with the transcription wrapping below when it does not fit.
    private struct WordLine: Block {
        let label: String
        let labelAttrs: [NSAttributedString.Key: Any]
        let labelBg: UIColor
        let text: String
        let textAttrs: [NSAttributedString.Key: Any]
        let tr: String?
        let trAttrs: [NSAttributedString.Key: Any]
        let gap: CGFloat
        let labelPadX: CGFloat
        let labelPadY: CGFloat
        let labelM: Metrics
        let labelH: CGFloat
        let labelW: CGFloat
        let textW: CGFloat
        let trW: CGFloat
        let oneLine: Bool
        let ascent: CGFloat
        let descent: CGFloat
        let line1H: CGFloat
        let line2Gap: CGFloat
        let line2H: CGFloat
        let trM: Metrics
        let height: CGFloat

        init(label: String, labelAttrs: [NSAttributedString.Key: Any], labelBg: UIColor, text: String, textAttrs: [NSAttributedString.Key: Any],
             tr: String?, trAttrs: [NSAttributedString.Key: Any], gap: CGFloat, maxWidth: CGFloat, s: CGFloat) {
            self.label = label
            self.labelAttrs = labelAttrs
            self.labelBg = labelBg
            self.text = text
            self.textAttrs = textAttrs
            self.tr = tr
            self.trAttrs = trAttrs
            self.gap = gap
            labelPadX = 14 * s
            labelPadY = 8 * s
            labelM = Metrics(labelAttrs)
            labelH = labelM.height + 2 * labelPadY
            labelW = CardRenderer.measure(label, labelAttrs) + 2 * labelPadX
            textW = CardRenderer.measure(text, textAttrs)
            trW = tr.map { CardRenderer.measure($0, trAttrs) } ?? 0
            oneLine = tr == nil || labelW + gap + textW + gap + trW <= maxWidth
            let textM = Metrics(textAttrs)
            trM = Metrics(trAttrs)
            ascent = max(textM.ascent, oneLine && tr != nil ? trM.ascent : 0)
            descent = max(textM.descent, oneLine && tr != nil ? trM.descent : 0)
            line1H = max(ascent + descent, labelH)
            line2Gap = 6 * s
            line2H = (!oneLine && tr != nil) ? trM.height + line2Gap : 0
            height = line1H + line2H
        }

        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            let total = labelW + gap + textW + ((oneLine && tr != nil) ? gap + trW : 0)
            var x = cx - total / 2
            let pillTop = y + (line1H - labelH) / 2
            ctx.setFillColor(labelBg.cgColor)
            ctx.addPath(UIBezierPath(roundedRect: CGRect(x: x, y: pillTop, width: labelW, height: labelH), cornerRadius: labelH / 2).cgPath)
            ctx.fillPath()
            CardRenderer.drawText(label, labelAttrs, x: x + labelPadX, baseline: pillTop + labelPadY + labelM.ascent, ascent: labelM.ascent)
            x += labelW + gap
            let baseline = y + (line1H - (ascent + descent)) / 2 + ascent
            CardRenderer.drawText(text, textAttrs, x: x, baseline: baseline, ascent: Metrics(textAttrs).ascent)
            if let tr {
                if oneLine {
                    CardRenderer.drawText(tr, trAttrs, x: x + textW + gap, baseline: baseline, ascent: trM.ascent)
                } else {
                    CardRenderer.drawText(tr, trAttrs, x: cx - trW / 2, baseline: y + line1H + line2Gap + trM.ascent, ascent: trM.ascent)
                }
            }
        }
    }

    private struct Paragraph: Block {
        let attributed: NSAttributedString
        let width: CGFloat
        let height: CGFloat
        init(_ text: String, _ attrs: [NSAttributedString.Key: Any], width: CGFloat, lineSpacing: CGFloat) {
            let style = NSMutableParagraphStyle()
            style.alignment = .center
            style.lineHeightMultiple = lineSpacing
            style.lineBreakMode = .byTruncatingTail
            var a = attrs
            a[.paragraphStyle] = style
            attributed = NSAttributedString(string: text, attributes: a)
            self.width = width
            let font = (attrs[.font] as? UIFont) ?? UIFont.systemFont(ofSize: 16)
            let maxH = ceil(font.lineHeight * lineSpacing * 3)
            let rect = attributed.boundingRect(with: CGSize(width: width, height: .greatestFiniteMagnitude), options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
            height = min(ceil(rect.height), maxH)
        }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            attributed.draw(with: CGRect(x: cx - width / 2, y: y, width: width, height: height), options: [.usesLineFragmentOrigin, .usesFontLeading, .truncatesLastVisibleLine], context: nil)
        }
    }

    private struct Divider: Block {
        let width: CGFloat
        let height: CGFloat
        let color: UIColor
        init(width: CGFloat, thickness: CGFloat, color: UIColor) {
            self.width = width
            height = thickness
            self.color = color
        }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            ctx.setFillColor(color.cgColor)
            ctx.fill(CGRect(x: cx - width / 2, y: y, width: width, height: height))
        }
    }

    private struct ImageTile: Block {
        let image: UIImage
        let size: CGFloat
        let radius: CGFloat
        let tileColor: UIColor
        var height: CGFloat { size }
        func draw(cx: CGFloat, y: CGFloat, ctx: CGContext) {
            ctx.setFillColor(tileColor.cgColor)
            ctx.addPath(UIBezierPath(roundedRect: CGRect(x: cx - size / 2, y: y, width: size, height: size), cornerRadius: radius).cgPath)
            ctx.fillPath()
            let inner = size * 0.64
            image.draw(in: CGRect(x: cx - inner / 2, y: y + (size - inner) / 2, width: inner, height: inner))
        }
    }
}

private protocol Block {
    var height: CGFloat { get }
    func draw(cx: CGFloat, y: CGFloat, ctx: CGContext)
}

/// Part-of-speech keys from the content, localised through the string catalog.
enum PosNames {
    static let known = ["noun", "verb", "adjective", "adverb", "interjection", "phrase", "pronoun", "preposition", "numeral", "conjunction"]

    static func localized(_ pos: String) -> String {
        known.contains(pos) ? L10n.t("pos.\(pos)") : pos
    }
}
